substrate/primitives/runtime/src/proving_trie/base2.rs

```diff
@@ -0,0 +1,306 @@
+where
+	Hashing: sp_core::Hasher,
+{
+	// Deduplicated and flattened list of key value pairs.
+	db: BTreeMap<Key, Value>,
+	root: Hashing::Out,
+	_phantom: core::marker::PhantomData<(Key, Value)>,
+}
+
+impl<Hashing, Key, Value> ProvingTrie<Hashing, Key, Value> for BasicProvingTrie<Hashing, Key, Value>
+where
+	Hashing: sp_core::Hasher,
+	Hashing::Out: Encode + Decode,
+	Key: Encode + Decode + Ord,
+	Value: Encode + Decode + Clone,
+{
+	/// Create a new instance of a `ProvingTrie` using an iterator of key/value pairs.
+	fn generate_for<I>(items: I) -> Result<Self, DispatchError>
+	where
+		I: IntoIterator<Item = (Key, Value)>,
+	{
+		let mut db = BTreeMap::default();
+		for (key, value) in items.into_iter() {
+			db.insert(key, value);
+		}
+		let root = merkle_root::<Hashing, _>(db.iter().map(|item| item.encode()));
+		Ok(Self { db, root, _phantom: Default::default() })
+	}
+
+	/// Access the underlying trie root.
+	fn root(&self) -> &Hashing::Out {
+		&self.root
+	}
+
+	/// Query a value contained within the current trie. Returns `None` if the
+	/// nodes within the current `db` are insufficient to query the item.
+	fn query(&self, key: &Key) -> Option<Value> {
+		self.db.get(&key).cloned()
+	}
+
+	/// Create a compact merkle proof needed to prove a single key and its value are in the trie.
+	/// Returns an error if the nodes within the current `db` are insufficient to create a proof.
+	fn create_proof(&self, key: &Key) -> Result<Vec<u8>, DispatchError> {
+		let mut encoded = Vec::with_capacity(self.db.len());
+		let mut found_index = None;
+
+		// Find the index of our key, and encode the (key, value) pair.
+		for (i, (k, v)) in self.db.iter().enumerate() {
+			// If we found the key we are looking for, save it.
+			if k == key {
+				found_index = Some(i);
+			}
+
+			encoded.push((k, v).encode());
+		}
+
+		let index = found_index.ok_or(TrieError::IncompleteDatabase)?;
+		let proof = merkle_proof::<Hashing, Vec<Vec<u8>>, Vec<u8>>(encoded, index as u32);
+		Ok(proof.encode())
+	}
+
+	/// Verify the existence of `key` and `value` in a given trie root and proof.
+	fn verify_proof(
+		root: &Hashing::Out,
+		proof: &[u8],
+		key: &Key,
+		value: &Value,
+	) -> Result<(), DispatchError> {
+		verify_proof::<Hashing, Key, Value>(root, proof, key, value)
+	}
+}
+
+impl<Hashing, Key, Value> ProofToHashes for BasicProvingTrie<Hashing, Key, Value>
+where
+	Hashing: sp_core::Hasher,
+	Hashing::Out: MaxEncodedLen + Decode,
+	Key: Decode,
+	Value: Decode,
+{
+	// This base 2 merkle trie includes the number of items in the trie, which we can directly use
+	// to figure out the depth of the trie.
+	fn proof_to_hashes(proof: &[u8]) -> Result<u32, DispatchError> {
+		let decoded_proof: MerkleProof<Hashing::Out, Vec<u8>> =
+			Decode::decode(&mut &proof[..]).map_err(|_| TrieError::IncompleteProof)?;
+		// Base 2 trie should have depth log2(n).
+		let depth = log2_rounded_up(decoded_proof.number_of_leaves);
+		Ok(depth)
```

Comment:
Maybe name it `proof_to_depth` then? I assume it will use the dept then to calculate the number of hashes.

Reply:
But depth is not the relevant number for benchmarking. Hashes are. It happens that depth and hashes are the same here, but it needs not be for strange merkle tries